ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala

```diff
@@ -251,6 +258,34 @@ object KeyValueConsumption {
     }
   }
 
+  /** Checks that all divulged contract ids in the `blindingInfo` have
+    * an associated contract instance in the `divulgedContracts` list.
+    *
+    * NOTE: It is not a hard fault for the check to fail,
+    * but missing contracts can cause lookup failures during command interpretation if used.
+    * This can happen if the participant is hydrated from ledgers populated with versions of
+    * [[com.daml.ledger.participant.state.kvutils.committer.transaction.TransactionCommitter]]
+    * predating the introduction of the `divulgedContracts` field.
+    */
+  private def checkDivulgedContractInstances(
+      divulgedContracts: List[DivulgedContract],
+      blindingInfo: DamlTransactionBlindingInfo,
+  ): Unit = {
+    val divulgencesFromBlindingInfo = blindingInfo.getDivulgencesList.asScala
+    if (divulgencesFromBlindingInfo.nonEmpty) {
+      val divulgedContractsSet = divulgedContracts.view.map(_.contractId.coid).toSet
+      val nonPresent = divulgencesFromBlindingInfo.filterNot(divulgence =>
+        divulgedContractsSet(divulgence.getContractId)
+      )
+      logger.debug(
+        s"Missing divulged contracts for contract ids: ${nonPresent.map(_.getContractId).mkString(",")}."
+      )
+      logger.warn(
+        "Missing divulged contracts in the decoded transaction entry. Command submissions using divulged contracts might fail."
```

Comment:
Add a reference to the transaction-id of the entry, so that it can be investigated easily.

Reply:
Addressed in the refactoring.